ext/nio4r/org/nio4r/ByteBuffer.java

```diff
@@ -0,0 +1,283 @@
+    }
+
+    @JRubyMethod(name = "remaining")
+    public IRubyObject remainingPositions(ThreadContext context) {
+        Ruby ruby = context.getRuntime();
+        int count = byteBuffer.remaining();
+        return RubyNumeric.int2fix(ruby, count);
+    }
+
+    @JRubyMethod(name = "remaining?")
+    public IRubyObject hasRemaining(ThreadContext context) {
+        if (byteBuffer.hasRemaining())
+            return context.getRuntime().getTrue();
+
+        return context.getRuntime().getFalse();
+    }
+
+    @JRubyMethod(name = "offset?")
+    public IRubyObject getOffset(ThreadContext context) {
+        int offset = byteBuffer.arrayOffset();
+        return JavaUtil.convertJavaToRuby(context.getRuntime(), offset);
+    }
+
+    /**
+     * Check whether the two ByteBuffers are the same.
+     *
+     * @param context
+     * @param ob      : The RubyObject which needs to be check
+     * @return
+     */
+    @JRubyMethod(name = "equals?")
+    public IRubyObject equals(ThreadContext context, IRubyObject ob) {
+        boolean equal = this.byteBuffer.equals(((ByteBuffer) JavaUtil.convertRubyToJava(ob)).getBuffer());
+        if (equal)
+            context.getRuntime().getTrue();
+        return context.getRuntime().getFalse();
+    }
+
+    /**
+     * Flip capability provided by the java nio.ByteBuffer
+     * buf.put(magic);    // Prepend header
+     * in.read(buf);      // Read data into rest of buffer
+     * buf.flip();        // Flip buffer
+     * out.write(buf);    // Write header + data to channel
+     *
+     * @param context
+     * @return
+     */
+    @JRubyMethod
+    public IRubyObject flip(ThreadContext context) {
+        byteBuffer.flip();
+        return this;
+    }
+
+    /**
+     * Rewinds the buffer. Usage in java is like
+     * out.write(buf);    // Write remaining data
+     * buf.rewind();      // Rewind buffer
+     * buf.get(array);    // Copy data into array
+     *
+     * @param context
+     * @return
+     */
+    @JRubyMethod
+    public IRubyObject rewind(ThreadContext context) {
+        byteBuffer.rewind();
+        return this;
+    }
+
+    @JRubyMethod
+    public IRubyObject reset(ThreadContext context) {
+        byteBuffer.reset();
+        return this;
+    }
+
+    @JRubyMethod
+    public IRubyObject mark(ThreadContext context) {
+        byteBuffer.mark();
+        return this;
+    }
+
+    /**
+     * Removes all the content in the byteBuffer
+     *
+     * @param context
+     * @return
+     */
+    @JRubyMethod
+    public IRubyObject clear(ThreadContext context) {
+        byteBuffer.clear();
+        return this;
+    }
+
+    @JRubyMethod
+    public IRubyObject compact(ThreadContext context) {
+        byteBuffer.compact();
+        return this;
+    }
+
+    @JRubyMethod(name = "capacity")
+    public IRubyObject capacity(ThreadContext context) {
+        int cap = byteBuffer.capacity();
+        return JavaUtil.convertJavaToRuby(context.getRuntime(), cap);
```

Comment:
with primitive `int`, `long` values one could do `context.runtime.newFixnum(int)` directly.
... its really a minor issue in this case but it will be a bit faster and avoids boxing.